ouroboros-network/src/Ouroboros/Network/PeerSelection/Churn.hs

```diff
@@ -61,180 +77,363 @@ peerChurnGovernor tracer deadlineChurnInterval bulkChurnInterval psOverallTimeou
-    increaseActiveBigLedgerPeers :: ChurnMode -> STM m ()
-    increaseActiveBigLedgerPeers mode =
-        modifyTVar peerSelectionVar (\targets -> targets {
-          -- TODO: when chain-skipping will be implemented and chain-sync client
-          -- will take into account big ledger peers, we don't need pattern
-          -- match on the churn mode, but use
-          -- `targetNumberOfActiveBigLedgerPeers` (issue #4609).
-          targetNumberOfActiveBigLedgerPeers =
-              case mode of
-                   ChurnModeNormal ->
-                      targetNumberOfActiveBigLedgerPeers base
-                   ChurnModeBulkSync ->
-                      min 1 (targetNumberOfActiveBigLedgerPeers base)
-        })
-
-    decreaseActiveBigLedgerPeers :: ChurnMode -> STM m ()
-    decreaseActiveBigLedgerPeers mode =
-        modifyTVar peerSelectionVar (\targets -> targets {
-          targetNumberOfActiveBigLedgerPeers =
-              case mode of
-                   ChurnModeNormal ->
-                       decrease $ targetNumberOfActiveBigLedgerPeers base
-                   ChurnModeBulkSync ->
-                       min 1 (targetNumberOfActiveBigLedgerPeers base)
-        })
-
-    decreaseEstablishedBigLedgerPeers :: STM m ()
-    decreaseEstablishedBigLedgerPeers =
-        modifyTVar peerSelectionVar (\targets -> targets {
-            targetNumberOfEstablishedBigLedgerPeers =
-              decrease (targetNumberOfEstablishedBigLedgerPeers base -
-                        targetNumberOfActiveBigLedgerPeers base)
-              + targetNumberOfActiveBigLedgerPeers base
-        })
-
-
-
-    go :: StdGen -> m Void
-    go !rng = do
+    increaseActivePeers :: ChurnMode
+                        -> ModifyPeerSelectionTargets
+    increaseActivePeers
+      mode targets
+      =
+      targets {
+        targetNumberOfActivePeers =
+          case mode of
+            ChurnModeNormal  ->
+              targetNumberOfActivePeers base
+            ChurnModeBulkSync ->
+              min 2 (targetNumberOfActivePeers base)
+      }
+
+    checkActivePeersIncreased :: CheckPeerSelectionCounters
+    checkActivePeersIncreased
+      PeerSelectionCounters { numberOfActivePeers }
+      PeerSelectionTargets { targetNumberOfActivePeers }
+      =
+      numberOfActivePeers >= targetNumberOfActivePeers
+
+
+    decreaseActivePeers :: ChurnMode -> ModifyPeerSelectionTargets
+    decreaseActivePeers mode targets =
+      targets {
+        targetNumberOfActivePeers =
+          case mode of
+            ChurnModeNormal ->
+              decrease $ targetNumberOfActivePeers base
+            ChurnModeBulkSync ->
+              min 1 (targetNumberOfActivePeers base - 1)
+      }
+
+    checkActivePeersDecreased :: CheckPeerSelectionCounters
+    checkActivePeersDecreased
+      PeerSelectionCounters { numberOfActivePeers }
+      PeerSelectionTargets { targetNumberOfActivePeers }
+      =
+      numberOfActivePeers <= targetNumberOfActivePeers
```

Comment:
Perhaps the next step, ie. to increase number of active peers, should take into account that hot demotions can take a long time, and increase the target not to targetNumberOfActivePeers, but to targetNumberOfActivePeers + numberOfHotDemotions. After a while, the demoted peers will fade, and we will eventually end up with targetNumberOfActivePeers. Does this make sense?

Reply:
Once peers are already being demoted we are guaranteed to pick the missing number of peers at some point.  The demotion action has a timeout itself, so it can't take forever.  This timeout is much shorter than the churn interval, so I think  it should be fine as is.

We are also exposing additional counters, which should give us better clue how it works in a running system.  Let's see if we can spot something strange.

Reply:
> I think that the check should be changed to:
> numberOfActivePeers - numberOfDemoteHot <= targetNumberOfActivePeers
> where numberOfDemoteHot = Set.size $ inProgressDemoteHot and should be added to PeerSelectionCounters.
> This means that we can move on as soon as the governor has picked some peers to demote.

I oppose this change.  It's hard to make a churn counter which can make sense: if we optimistically trace that we decreased the peers in progress, then the counter which logs how many peers we promoted will be smaller by that number, which will be confusing.

Churn can take some time, so I don't mind to have a 300s timeout on decreasing the number of active peers.  If churn takes 10 minutes every hour that should be fine.

Reply:
This is how churn graphs looks without this change, like it was done initially:
![image](https://github.com/IntersectMBO/ouroboros-network/assets/698209/9aac8c48-912f-447c-844f-b1a3ea358c9a)